### ISSUE TITLE
Cannot iterate over null (prepare.py)

### ISSUE THREAD

**Issue by sq-brootz** [duo-labs/cloudmapper#205]
Similar to issue #166.   Still getting this error with a new clone of the repo but in the prepare instead of collect.

Command run:
```
python3 cloudmapper.py prepare --account my_account
```

error:
```
Traceback (most recent call last):
File "cloudmapper.py", line 69, in <module>
      main()
File "cloudmapper.py", line 64, in main
      commands[command].run(arguments)
File "/Users/ehansen/projects/cloudmapper/commands/prepare.py", line 419, in run
      prepare(account, config, outputfilter)
File "/Users/ehansen/projects/cloudmapper/commands/prepare.py", line 345, in prepare
      cytoscape_json = build_data_structure(account, config, outputfilter)
File "/Users/ehansen/projects/cloudmapper/commands/prepare.py", line 306, in build_data_structure
      for c, reasons in get_connections(cidrs, vpc, outputfilter).items():
File "/Users/ehansen/projects/cloudmapper/commands/prepare.py", line 145, in get_connections
      for sg in instance.security_groups():
File "/Users/ehansen/projects/cloudmapper/shared/nodes.py", line 320, in security_groups
      return pyjq.all('.SecurityGroups[]', self._json_blob)
File "/usr/local/lib/python3.7/site-packages/pyjq.py", line 49, in all
      return compile(script, vars, library_paths).all(_get_value(value, url, opener))
File "_pyjq.pyx", line 209, in _pyjq.Script.all (_pyjq.c:2561)
_ . pyjq.ScriptRuntimeError: Cannot iterate over null (null)
```

**Comment by auroan** [comment]
I get a similar problem;

~/cloudmapper$ python3 cloudmapper.py prepare --account CDL --regions eu-central-1
Building data for account <removed>
Traceback (most recent call last):
  File "cloudmapper.py", line 69, in <module>
    main()
  File "cloudmapper.py", line 64, in main
    commands[command].run(arguments)
  File "/home/danny/cloudmapper/commands/prepare.py", line 419, in run
    prepare(account, config, outputfilter)
  File "/home/danny/cloudmapper/commands/prepare.py", line 345, in prepare
    cytoscape_json = build_data_structure(account, config, outputfilter)
  File "/home/danny/cloudmapper/commands/prepare.py", line 240, in build_data_structure
    outputfilter["collapse_asgs"])
  File "/home/danny/cloudmapper/shared/nodes.py", line 285, in __init__
    autoscaling_name = pyjq.all('.Tags[] | select(.Key == "aws:autoscaling:groupName") | .Value', json_blob)
  File "/home/danny/.local/lib/python3.6/site-packages/pyjq.py", line 49, in all
    return compile(script, vars, library_paths).all(_get_value(value, url, opener))
  File "_pyjq.pyx", line 209, in _pyjq.Script.all (_pyjq.c:2561)
_pyjq.ScriptRuntimeError: Cannot iterate over null (null)

**Comment by auroan** [comment]
Fixed the issue I was having by adding "?" to two sections in node.py

285c285
<             autoscaling_name = pyjq.all('.Tags[] | select(.Key == "aws:autoscaling:groupName") | .Value', json_blob)
---
>             autoscaling_name = pyjq.all('.Tags[]? | select(.Key == "aws:autoscaling:groupName") | .Value', json_blob)
289c289
<             collapse_by_tag_value = pyjq.all('.Tags[] | select(.Key == "{}") | .Value'.format(collapse_by_tag), json_blob)
---
>             collapse_by_tag_value = pyjq.all('.Tags[]? | select(.Key == "{}") | .Value'.format(collapse_by_tag), json_blob)

**Comment by sq-brootz** [comment]
Thanks, @auroan.  In addition to those lines I also had to do the same to the following:

nodes.py (orig)
```
319    def security_groups(self):
320       return pyjq.all('.SecurityGroups[]', self._json_blob)
```

nodes.py (changed)
```
319    def security_groups(self):
320       return pyjq.all('.SecurityGroups[]?', self._json_blob)
```

Seems to be working correctly now!

**Comment by ops-ninja** [comment]
I got this while running collect.py as well. I worked around it by preparing the incoming strings for nulls before calling pyjq. e.g.
186        pyjq_parse_string = pyjq_parse_string.replace("[]","[]?")
As well as the above noted change on line 320 of nodes.py e.g
320       return pyjq.all('.SecurityGroups[]?', self._json_blob)

**Comment by rsiripan** [comment]
> Fixed the issue I was having by adding "?" to two sections in node.py
> 
> Orig
> 
> autoscaling_name = pyjq.all('.Tags[] | select(.Key == "aws:autoscaling:groupName") | .Value', json_blob)
> 
> Updated
> 
> autoscaling_name = pyjq.all('.Tags[]**?** | select(.Key == "aws:autoscaling:groupName") | .Value', json_blob)
> 
> Orig
> 
> collapse_by_tag_value = pyjq.all('.Tags[] | select(.Key == "{}") | .Value'.format(collapse_by_tag), json_blob)
> Updated
> 
> collapse_by_tag_value = pyjq.all('.Tags[]**?** | select(.Key == "{}") | .Value'.format(collapse_by_tag), json_blob)

**Comment by rsiripan** [comment]
I have updated both @auroan and @sq-brootz has mentioned but still got this error message, any clue?


 _Traceback (most recent call last):
  File "cloudmapper.py", line 70, in <module>
    main()
  File "cloudmapper.py", line 54, in main
    module = importlib.import_module(full_package_name)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/ray/gits/cloudmapper/commands/amis.py", line 6, in <module>
    import pyjq
ImportError: No module named pyjq_

**Comment by xlash** [comment]
The fix is easy, this should be merge back.

**Comment by chrisswanda** [comment]
I fixed mine by adding the "?" to:

        return pyjq.all('.SecurityGroups[]?.GroupId', self._json_blob)
            autoscaling_name = pyjq.all('.Tags[]? | select(.Key == "aws:autoscaling:groupName") | .Value', json_blob)
            collapse_by_tag_value = pyjq.all('.Tags[]? | select(.Key == "{}") | .Value'.format(collapse_by_tag), json_blob)

**Comment by evilpete** [comment]
The issue with 
`pyjq.all('.SecurityGroups[]?', self._json_blob)`
is that elbv2 data does not have the field  SecurityGroups  if there are  no security groups attached to the ELB.

**Comment by 0xdabbad00** [comment]
@evilpete Thank you for the explanation! Do you know why someone would have a ELB without Security Groups?  Has it just not yet been setup or is there a reason to do that?

**Comment by evilpete** [comment]
Let's just say I discovered this at a client site I'm cleaning up..

**Comment by 0xdabbad00** [comment]
Thanks all. Sorry this bug had been around for so long. It didn't make sense to me until @evilpete's comment.  Should be resolved in #432

**Comment by azarboon** [comment]
Still I have the same issue. Also I' wondering what should i put for "cidrs" in config.json? Can I leave it empty or what?
